contracts/Vault.vy

```diff
@@ -553,36 +556,40 @@ def setWithdrawalQueue(queue: address[MAXIMUM_STRATEGIES]):
     """
     assert msg.sender in [self.management, self.governance]
 
-    # HACK: Temporary until Vyper adds support for Dynamic arrays
-    old_queue: address[MAXIMUM_STRATEGIES] = empty(address[MAXIMUM_STRATEGIES])
+    set: address[SET_SIZE] = empty(address[SET_SIZE])
     for i in range(MAXIMUM_STRATEGIES):
-        old_queue[i] = self.withdrawalQueue[i] 
         if queue[i] == ZERO_ADDRESS:
             # NOTE: Cannot use this method to remove entries from the queue
-            assert old_queue[i] == ZERO_ADDRESS
+            assert self.withdrawalQueue[i] == ZERO_ADDRESS
             break
         # NOTE: Cannot use this method to add more entries to the queue
-        assert old_queue[i] != ZERO_ADDRESS
+        assert self.withdrawalQueue[i] != ZERO_ADDRESS
 
         assert self.strategies[queue[i]].activation > 0
 
-        existsInOldQueue: bool = False
-        for j in range(MAXIMUM_STRATEGIES):
-            if queue[j] == ZERO_ADDRESS:
-                existsInOldQueue = True
+        # hash will have hash of the address in uint256.
+        # `mask = SET_SIZE - 1` will be used to get key for the set for indexing
+        hash: uint256 = convert(queue[i], uint256)
+        key: uint256 = bitwise_and(hash, SET_SIZE - 1)  # Key is first `SET_SIZE` bits of address
+        # Most of the times following for loop only run once which is making it highly gas efficient 
+        # but in the worst case of key collision it will run linearly and find first empty slot.
+        for m in range(key, key + SET_SIZE):
+            if m >= SET_SIZE:
+                for n in range(SET_SIZE):
+                    # Note: This only means SET_SIZE is lower than MAXIMUM_STRATEGIES
+                    assert n < key
+                    assert set[n] != queue[i]
+                    if set[n] == ZERO_ADDRESS:
+                        set[n] = queue[i]
+                        break
+                break
```

Comment:
Doesn't this mean it exhausted the set available space without finding a place for the entry?

Reply:
Nope. it will only get triggered when it founds an empty space within the set. other cases will raise an error due to assertions.

Reply:
This is not strictly true actually, the `break` internal to the `if` statement above it should trigger if it finds space in the set, this line needs to change to be a `raise` to account for the scenario where no space is found at all
```suggestion
                # NOTE: should not revert if SET_SIZE is lower than MAXIMUM_STRATEGIES
                raise UNREACHABLE  # dev: no space in set found
```

Reply:
> Ex: `key = 30` and both 30 and 31 is filled. Now we start with 0th slot inside the `if m >= SET_SIZE` part, and start looping from start to find empty space if we find the same element at any place we flag it as duplicate and raise an error. if we find an empty slot we put the strategy address there and break the **inner loop**. control goes to the `if block` outside the inner loop which is inside another loop which you need to break and register the strategy in withdrawal queue.
> 
> If you raise an error there a valid transaction would revert.

Ah, I see the difference now.

The change would be this:
```suggestion
                    # NOTE: should not revert if SET_SIZE is lower than MAXIMUM_STRATEGIES
                    raise UNREACHABLE  # dev: no space in set found

                # NOTE: break out of top loop
                break
```
With this change, removing `assert n < key` above would be valid as that invariant check happens here.

---

Alternatively, one could just `m %= SET_SIZE` on the first line to re-align the set index to be within the loop's boundaries